# Statistics

**Statistics** is a branch of mathematics dealing with the **collection, analysis, interpretation, presentation, and organization of data**

A statistic (singular) or sample statistic is a** single measure of some attribute** of a sample (e.g. its arithmetic mean value).

# Descriptive Statistics

**Descriptive Statistics** summaries a dataset, which helps in **gaining insights**, and making inferences about a dataset.

In general, we compute statistical measures of **one or more samples**, related to a **population**, and draw conclusions about the population.

**Descriptive Statistics** involves estimating **centrality measures** and **measures of dispersion**.

# Centrality Measures

Centrality measures determine the **center** of a dataset.

The three major centrality measures are: **mean**, **median**, and **mode**.

## Mean

Mean is the **sum of all values divided by a total number of values**, of a data set.

**mean** function of **numpy** is used to compute mean marks obtained by **five** students in a subject **s1**.

In [118]:
import numpy as np

s1 = np.array([86, 47, 45, 47, 40])

print(np.mean(s1))

53.0


## Median

Median is the value that **separates** the given data set into **two halves**.

**median** function of **numpy** library can be used to compute median of a data set, as shown here,

In [119]:
import numpy as np

s1 = np.array([86, 47, 45, 47, 40])

print(np.median(s1))

47.0


## Mode

The value that occurs a **number of times** in a data set refers to **Mode** of the data set.

**mode** function of **scipy.stats** module can be used for computing mode of a given data set.

In [120]:
from scipy import stats

s1 = np.array([86, 47, 45, 47, 40])

print(stats.mode(s1))

ModeResult(mode=array([47]), count=array([2]))


In [121]:
from scipy import stats
print(stats.mode([8, 9, 8, 7, 9, 6, 7, 6]))

ModeResult(mode=array([6]), count=array([2]))


If there are **more than a single** value, then only the **first** one is returned by mode function.

## Measures of Dispersion

**Measures of Dispersion** provide **insights on the spread** of given dataset.

Major measures of dispersion are: **range, percentile, inter-quartile range, standard deviation, variance, skewness, and kurtosis**.

### Range

**Range** is the difference between **maximum** and **minimum** values of the dataset.

In [122]:
import numpy as np

s1 = np.array([86, 47, 45, 47, 40])

print(np.ptp(s1))

46


### Percentile

**Percentile** refers to a value, below which **lies given the percentage** of data points.

E.g., **45th** percentile refers to a value below which **45%** of data points are found.

**percentile** function of **numpy** can be used to compute a **single** or **multiple** percentiles.

In [123]:
import numpy as np

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])

print(np.percentile(s2, 45, interpolation='lower'))

65


### Quartiles

Three **Quartiles** namely, **Q1** **Q2** and **Q3**, split the entire dataset into **four equal parts**.

**Each part **contains **25%** of data.

In [124]:
import numpy as np

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])

print(np.percentile(s2, [25, 50, 75], interpolation='lower'))

[47 65 83]


### Inter Quartile Range (IQR)

**Inter quartile range **refers to difference between **third quartile (Q3)** and **first quartile (Q1).**

**iqr** method from **scipy.stats** can be used for calculating it.


In [125]:
import numpy as np
from scipy import stats

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])

print(stats.iqr(s2, rng=(25, 75), interpolation='lower'))

36


## Variance and Standard Deviation

### Variance

**Variance** is defined as the **average of squared differences**, of each data point from dataset's mean.

### Standard Deviation

**Standard Deviation** is **square root** of **variance**.

**var** and **std** functions of **numpy** can be used for computing **variance** and **standard deviation** respectively.

By default, the functions assume that the dataset represents entire population.

To represent a sample, derived from a population, **ddof** parameter is set to 1.

In [126]:
import numpy as np

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])


print(np.var(s2))
print(np.std(s2))

# s2 representing a population's sample
print(np.var(s2, ddof=1))
print(np.std(s2, ddof=1))

454.21000000000004
21.312203077110542
504.6777777777778
22.46503455990615


## Skewness

**skewness** determines whether the majority of data points are present on **one side of the distribution**.

A **positive** value represents right **skewed distribution**; a **negative** value represents **left skewed one**, **zero** represent **unskewed distribution**.

In [127]:
import numpy as np
from scipy import stats

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])

print(stats.skew(s2))

0.04321019390325423


## Kurtosis

**Kurtosis** indicates how much of data is **concentrated** around **mean** or **shape** of the **probability distribution**.

It can be **estimated** using **kurtosis** function of **scipy.stats** module.

By default, it uses **Fisher**’s definition. This can be changed to **Pearson** by setting **fisher parameter** to **False**.

In [128]:
import numpy as np
from scipy import stats

s2 = np.array([86, 47, 45, 47, 40, 97, 98, 75, 65, 83])

print(stats.kurtosis(s2))

-1.5694354898634155


## HandsOn

Compute the following statistical parameters, and display in separate lines, for the sample data set s = [26, 15, 8, 44, 26, 13, 38, 24, 17, 29].

Mean,

Median,

Mode,

25th and 75th percentile,

Inter quartile range,

Skewness,

Kurtosis.

***Hint***: Import **stats** from **scipy** and Set **interpolation parameter value** to **lower** for computing** inter quartile range**.

In [129]:
import numpy as np
from scipy import stats
s = np.array([26, 15, 8, 44, 26, 13, 38, 24, 17, 29])
print(np.mean(s))
print(np.median(s))
print(stats.mode(s))
print(np.percentile(s,[25,75]))
print(stats.iqr(s, rng=(25, 75), interpolation='lower'))
print(stats.skew(s))
print(stats.kurtosis(s))


24.0
25.0
ModeResult(mode=array([26]), count=array([2]))
[15.5  28.25]
11
0.3622439411783622
-0.7666571612775241


# Random Numbers

A **random number** is a number, **chosen by chance** from a distribution.

Python provides a lot of modules, which deal with random numbers.


*   **random** module of python standard library.
*   **random** module of **Numpy**
*   **stats** module of **Scipy**





## Numpy's Random Module


*   **random** module of **numpy** has utilities, which generate arrays of random numbers.
*   E.g.: **rand** function generates uniformly distributed numbers from **range [0, 1]**
*   **rand** function with no arguments generate a **single** random value.
*   By passing arguments, it generates a random array of **specified size**.





In [130]:
import numpy as np

print(np.random.rand())

# generates a 2*3 array
print(np.random.rand(2,3))

0.38637572509456086
[[0.6110893  0.52316933 0.3250568 ]
 [0.92979042 0.84820141 0.84112295]]


## Random Sampling

In statistics, you select items **randomly** from a **population**, either **with or without** a **replacement**.

This can be achieved with **choice** method as shown below. Two items are selected **randomly**, **without replacement**.

In [131]:
import numpy as np

print(np.random.choice([11, 22, 33], 2, replace=False))

[11 33]


## Random Seeding

**Seed** is an important concept when it comes to **reproducibility**. If you are working with random numbers and you would want to peers to validate your results, i.e., they should also get the **same random sequence** as you did, you can set the seed to a particular value and send the **seed value** to your **peers**.



*   **seed** is a number that sets the initial state of random number generator.
*   Setting a seed, helps in generating the** same sequence** of random numbers, **repeatedly**.
*   **seed** method of a random module can be used to set a seed as shown in below example.





In [132]:
import numpy as np

np.random.seed(100)
print(np.random.rand())

np.random.seed(100)
print(np.random.rand())

0.5434049417909654
0.5434049417909654


## Random Variables



*   In probability theory, the **set of all possible **outcomes of a** random experiment **is known as **sample space**.
*   Probabilities of all outcomes of the experiment define the probability distribution.
*   A **random variable** is a variable that takes **real numbers or integers** and map each value to one of the outcomes of sample space.
*   E.g.: In an experiment of tossing a coin, the sample space is **{'Head', 'Tail'}** and a possible random variable takes the value **0 for head** and **1 for the tail**.



## Probability Distributions


*   There are **two** types of probability distributions namely **discrete** and **continuous** that take **integer and real values**, respectively.
*   **scipy.stats** module provides classes that represent random variables, corresponding to a large number of probability distributions.
*   E.g: the class **norm** represent normal **continuous** random variable, and **binom** represent binomial **discrete** random variable.





## Random Distributions

**scipy.stats** module provide a lot of methods for created **discrete** and **continuous** random variables.

Commonly used methods are described below.


*   **pdf / pmf** : Probability distribution function (continuous) or probability mass function (discrete).
*   **cdf** : Cumulative distribution function.
*   **sf** : Survival function (1 – cdf).
*   **rvs** : Creating random samples from a distribution.

The following example defines a normal continuous random variable of **mean 1.0** and **std 2.5**.

It also **estimates probabilities** and **cumulative probabilities** at **-1, 0 and 1**.

The example also generates **six** random numbers from defined normal distribution.



In [133]:
from scipy import stats

x = stats.norm(loc=1.0, scale=2.5)

print(x.pdf([-1, 0, 1]))

print(x.cdf([-1, 0, 1]))

print(x.rvs((2,3)))

[0.11587662 0.14730806 0.15957691]
[0.2118554  0.34457826 0.5       ]
[[-0.40409689 -3.12269376  1.88668613]
 [-0.96516083  0.42031951  1.5199392 ]]


## HandsOn

Create a normal distribution with **mean 32** and **standard deviation 4.5**.

Set the random **seed** to **1**.

Create a random sample of **100** elements, from defined distribution.

Compute the absolute difference between the sample **mean**, and the distribution **mean**.

***Hint***: Use functions available in **numpy** and **scipy**.

In [134]:
import numpy as np

from scipy import stats

np.random.seed(1)

x = stats.norm(loc=32.0, scale=4.5)

sample=x.rvs(100)

print(sample.mean()-x.mean())

0.2726228343406518


Simulate a random experiment of tossing a coin **10000** times and determine the **count of Heads**.

***Hint***: Define a binomial distribution with **n = 1** and **p = 0.5**.

Use **binom** function from **scipy.stats**.

Set the random **seed** to **1**.

Draw a sample of **10000** elements from defined distribution. Assume the values **0** and **1** represent **Heads** and **Tails** respectively.

Count the **number of heads** and **display it**. Make used of '**bincount**' method, available in '**numpy**'.

In [135]:
import numpy as np

from scipy import stats

np.random.seed(1)

x = stats.binom(n=1,p=0.5)

sample=x.rvs(10000)

print(np.bincount(sample)[0])

4990


# Hypothesis Testing Using scipy

## Hypothesis Testing

**Hypothesis Testing** is a methodology for evaluating if a **claim** is **acceptable** or **not**, based on data.

In a Hypothesis Testing, a **Null Hypothesis (Ho)** represents currently **accepted** the **state of knowledge**, and an **Alternative Hypothesis (Ha)** represents a new claim which **challenges** the currently** accepted state of knowledge**.

The **null hypothesis** and the **alternative hypothesis** are **mutually exclusive**.

## Steps Involved in Hypothesis Testing

The following steps are involved in a Hypothesis Testing:


*   Define the null hypothesis and the alternative hypothesis.
*   Select a test statistics whose probability distribution function can be found under the null hypothesis.
*   Collect data.
*   Compute the test statistics from the data and calculate its p-value under the null hypothesis.
*   Null hypothesis is rejected if the p-value is lower than predetermined significance value.


## Choosing Test Statistics

In hypothesis testing, selecting a test statistics is the most difficult part.

The methods used for performing t-test are shown below.

*   **stats.ttest_1samp**: Tests if the mean of a population is a given value.
*   **stats.ttest_ind**: Tests if the means of two independent samples are equal.
*   **stats.ttest_rel**: Tests if the means of two paired samples are equal.


### Example 1



*   Let's consider a common hypothesis: Mean of a population is equal to certain value.
*   In reality, we estimate mean and variance of a sample and calculate the test statistic.
*   If population variance is identified, then it is reasonable to consider that test statistic is normally distributed.
*   If population variance is unknown, sample variance is used, and test statistic follows t distribution.


Consider a normal population with **mean 0.8** and **standard deviation 0.5.**

Define the **null hypothesis** as **Mean** of the population is **1.0**.

Let's calculate t-statistic and p-value

In [136]:
from scipy import stats
import numpy as np

np.random.seed(1)
mu, sigma = 0.8, 0.5
X = stats.norm(mu, sigma)

# Deriving a sample
n = 100
X_sample = X.rvs(n)

# Computing test statistic
t, p = stats.ttest_1samp(X_sample, 1.0)
print(t, p)

-3.81532426532222 0.00023686273495632666


In previous example, the obtained **t-statistic** value is **-3.81** and **p-value** is **0.00023686**.

Since **p-value** is very **low** and less than the significance level **0.05**, you can **reject** the null hypothesis and infer that the mean of the population is **not 1**.

### Example 2

Let's consider another problem, where the **null hypothesis** states that the population means of **two random variables are equal**.

The below example derives two samples from different populations and verifies the claim that their population means are equal.

In [137]:
np.random.seed(2)
X1 = stats.norm(0.25, 1.0)
X2 = stats.norm(0.50, 1.0)

X1_sample = X1.rvs(100)
X2_sample = X2.rvs(100)

t, p = stats.ttest_ind(X1_sample, X2_sample)
print(t, p)

-3.157999488552839 0.0018371853479071398


In previous example, the obtained **t-statistic** value is **-3.157** and **p-value** is **0.0018**.

Since **p-value** is very low and less than the significance level **0.05**, you can **reject** the null hypothesis and state that population means of both **samples differ**.

## HandsOn

Consider the two independent samples s1 and s2 given below.

s1 = [45, 38, 52, 48, 25, 39, 51, 46, 55, 46]

s2 = [34, 22, 15, 27, 37, 41, 24, 19, 26, 36]

The samples represent **life satisfaction score**, computed through a **methodology**, of **older adults and younger adults** respectively.

Compute **t-statistic** for the above **two groups** and **display the t-score and the p value** in separate lines.

***Hint***: Use **ttest_ind** function available in **scipy**

In [138]:
import numpy as np
from scipy import stats
s1 = np.array([45, 38, 52, 48, 25, 39, 51, 46, 55, 46])
s2 = np.array([34, 22, 15, 27, 37, 41, 24, 19, 26, 36])
t,p=stats.ttest_ind(s1,s2)
print(t)
print(p)

4.257546665558161
0.0004736633119019225


A researcher noted the number of chocolate chips consumed by 10 rats, with and without electrical stimulation.

The data sets s1 represents consumption with stimulation and s2 without simulation.

s1 = [12, 7, 3, 11, 8, 5, 14, 7, 9, 10]

s2 = [8, 7, 4, 14, 6, 7, 12, 5, 5, 8]

Compute** t-statistic** for above considered two samples and **display** the **t-score and p-value** in separate lines

***Hint*** : Use **ttest_rel** function available in **scipy**

In [139]:
import numpy as np
from scipy import stats
s1 = np.array([12, 7, 3, 11, 8, 5, 14, 7, 9, 10])
s2 = np.array([8, 7, 4, 14, 6, 7, 12, 5, 5, 8])
t,p=stats.ttest_rel(s1,s2)
print(t)
print(p)

1.315587028960544
0.2208380130273219


# Introduction to Statistical Modelling

## Statistical Model

A **Statistical Model **is a mathematical equation, which explains the relationship between** dependent variables** (Y) and** independent variables** (X).

In general, a model is written as

**Y = f(X)**

However, in reality, an element of **uncertainty** is expected **due** to factors such as measurement **noise**. Hence the aforementioned equation can be rewritten as

**Y = f(X) + e**   '**e**' is residual error



*   Statistical Modelling deals with creating models that attempt to explain the data best.
*   The simplest model is a linear model represented as  **Y = B0 + B1*X + e**,  where the coefficients, **B0** and **B1** are the parameters of the model and **e** is normally distributed residual error.
*   A linear regression model assumes that residuals are independent and normally distributed.
*   The model is fitted to data using ordinary least squares approach.



## Linear Models

In most of the linear regression models, a dependent variable Y is written as :

*   a linear combination of the response variables X i.e  **Y = B0 + B1*X1 + ... + Bn*Xn**, or
*   functions of the response variables i.e  **Y = B0 + B1*X + B2*x^2 + ... + Bn*X^n**, or
*   models that have a linear component i.e  **Y = B0 + B1*sin(X1) + B2*cos(X2)**




## Non-Linear Models

*   Other than linear regression models, statistical modeling can be used to build non-linear models.
*   Errors of dependent variable follow a distribution other than normal distribution.
*   Examples of non-linear models are: **Binomial Regression and Poisson Regression.**
*   In most of the cases, the non-linear models are generalized to linear models.



## Design Matrices


*   In reality, you choose a model and fit the available data to it.
*   Once a model is chosen, design matrices **y** and **X** are constructed, and the regression problem is written, in matrix form, as **y = XB + e.**
    *   where **y** is the vector of **observations**, **X** is the vector of **dependent** variables, **B** is a vector of **coefficients**, and **e** is the **residual (error)**.
*   Thus obtained design matrices are passed as inputs to the chosen model.



## Statistical Modelling with StatsModels



*   **statsmodels** library supports several types of statistical models.
*   All of them follow a similar usage pattern.
*   A Statistical model is represented by a model class.
*   A model can be initiated with,
    *   given design matrices of dependent and independent variables, or with
    *   given patsy formula and data frame or dictionary-like object.





### Step 1: Creating a Model

An instance of a model class is created in either of the following ways.

**model = sm.MODEL(y, X)** or

**model = smf.model(patsy_formula, data)**

where **MODEL** and **model** refer to **model names** such as **OLS, GLM, ols, glm, etc**.

**Uppercase** names take **design matrices** as arguments, and **lowercase** names take **Patsy formulas** and **data frames** as arguments.

### Step 2: Fitting a Model

In order to fit the model with data, fit method is invoked on the created model, as shown below.

**result = model.fit()**

The **fit** method returns a result object, which has methods and attributes for further analysis.

### Step 3: Viewing Model Summary

The **summary** method of result object produces a summary text that describes the result of the fit, as shown below.

**print(result.summary())**

The displayed summary text varies for each statistical model and provides information of various statistical parameters.

### Step 4: Analyzing the Model Further

Other than viewing summary statistics, you can perform activities like,


*   Determining fitted values,
*   Predicting the dependent variable values for new independent variable values.
*   Checking if residuals of fitted models follow a normal distribution or not.



# Defining Statistical Models with patsy

## Constructing Design Matrices

The below example shows calculation of design matrices, **y**, and **X**, for the considered linear model  **Y = B0 + B1*X1 + B2*X2 + B3*X1*X2**

In [140]:
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
X = np.vstack([np.ones(5), x1, x2, x1*x2]).T

print(y)
print(X)

[1 2 3 4 5]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


Thus obtained design matrices **(y and X)** can be passed to **regression methods** for obtaining the **coefficient vector**.

## Design Matrices with patsy

**patsy**, a python library, allows defining a model in simpler easily.

It also constructs relevant design matrices, automatically, using** patsy.dmatrices** function.

**patsy.dmatrices** takes a formula (in string form) as a first argument, and a dictionary-like object with data arrays for the response variables as second arguments.

## Patsy Design Matrices

An example of using** patsy.dmatrices** function is shown below.

In [141]:
import patsy
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
data = {'y':y, 'x1':x1, 'x2':x2}

y, X = patsy.dmatrices('y ~ 1 + x1 + x2 + x1*x2', data)

print(y)
print(X)

[[1.]
 [2.]
 [3.]
 [4.]
 [5.]]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


## Understanding patsy Formulae

Let's understand few patsy formulae provided below.

**'y ~ x'** : **y** is linearly dependent on **x**. **~** symbol **separates** **dependent** variable from **independent** variable terms. It is also equivalent to '**y ~ 1 + x**'.

**'y ~ x1 + x2'** : **y** is a linear combination of **x1** and **x2**. **+** sign is used to denote the **union** of terms.

 **y ~ x1*x2 : x1*x2** is an interaction term that includes** all lower order terms**. Hence formula is equivalent to  **y ~ 1 + x1 + x2 + x1*x2**.
 
** 'y ~ np.log(x1)'**: Often **numpy** functions can be used to transform terms in the expression.

**'y ~ I(x1 + x2)'**: **I** is the **identify** function, used to** escape arithmetic expressions** and are evaluated.

**'y ~ C(x1)'**: Treats the variable **x1** as a **categorical** variable.

# Example Datasets


*   **statsmodels** contain few popular example datasets, which can be used to explore various utilities of the package.
*  The example data sets are available in **datasets** module.
*   Each dataset is associated with special variables like **SOURCE**, **DESCSHORT**, **DESCLONG**, that provide more info about the dataset.
*   A dataset can be loaded using **load** function, and its data can be accessed using **data** attribute in the form of **Numpy**'s recarray.



## Loading Example Datasets

The below example shows loading of popular **breast cancer dataset** and accessing its data as **numpy** array.

In [142]:
import statsmodels.api as sm

bc_cancer_set = sm.datasets.cancer

bc_cancer = bc_cancer_set.load()

bc_cancer_data = bc_cancer.data

print(type(bc_cancer_data))

<class 'numpy.recarray'>


The data can be accessed in from of pandas data frame, if it is loaded with function load_pandas.

## Loading R Datasets

**statsmodels** package provides access to many example datasets of **R**, listed at [R Datasets repository](http://vincentarelbundock.github.io/Rdatasets/datasets.html)

The below example loads **Icecream** dataset from **Ecdat** package.

In [143]:
import statsmodels.api as sm

icecream_data = sm.datasets.get_rdataset('Icecream', 'Ecdat')
data = icecream_data.data

print(icecream_data.data.shape)

(30, 4)


# Linear Regression with statsmodels

## Linear Regression Example

Let's understand how to fit a linear regression model for Icecream dataset available from R Data Repository.

In [144]:
import statsmodels.api as sm

icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")

icecream_data = icecream.data

print(icecream_data.columns)

Index(['cons', 'income', 'price', 'temp'], dtype='object')


The **icecream_data** dataset is a **pandas** data frame. It contains four variables: **cons (consumption), income, price, and temp (temperature)**.

### Model Creation

**Choosing a Model**

Initially, let's model consumption with price and temperature, as a linear model.

The patsy formula for assumed model is: **cons ~ price + temp**

**Creating a Model**

In [0]:
import statsmodels.formula.api as smf

linear_model1 = smf.ols('cons ~ price + temp', icecream_data)

**Fitting the Model**

In [0]:
linear_result1 = linear_model1.fit()

### Viewing Model Summary

Model summary can be viewed using the command

In [147]:
print(linear_result1.summary())

                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.633
Model:                            OLS   Adj. R-squared:                  0.606
Method:                 Least Squares   F-statistic:                     23.27
Date:                Tue, 19 Jun 2018   Prob (F-statistic):           1.34e-06
Time:                        12:04:07   Log-Likelihood:                 54.607
No. Observations:                  30   AIC:                            -103.2
Df Residuals:                      27   BIC:                            -99.01
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5966      0.258      2.309      0.0

### Analyzing the Model

The R-squared value of **0.633** suggests that model is not a proper fit.

The probability value of coefficient **price** is high, i.e., **0.141**. This accepts the null-hypothesis: the value of **price** coefficient is equal to zero.

Hence, the variable **price** does not affect **cons** variable.

### Model Recreation 1

Now let's create a new model by considered **income** and **temp** dependent variables.

In [0]:
linear_model2 = smf.ols('cons ~ income + temp', icecream_data)

linear_result2 = linear_model2.fit()


### Viewing Recreated Model Summary 1

In [149]:
print(linear_result2.summary())

                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.702
Model:                            OLS   Adj. R-squared:                  0.680
Method:                 Least Squares   F-statistic:                     31.81
Date:                Tue, 19 Jun 2018   Prob (F-statistic):           7.96e-08
Time:                        12:04:09   Log-Likelihood:                 57.742
No. Observations:                  30   AIC:                            -109.5
Df Residuals:                      27   BIC:                            -105.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.1132      0.108     -1.045      0.3

### Analyzing Model 1

R-squared value improved to **0.702**, suggesting this is a better model than previous one.

Probability values of **income** and **temp** are low, indicating they are highly significant to **cons**.

The probability value of **Intercept** is too **high**.

### Model Recreation 2


In [0]:
linear_model3 = smf.ols('cons ~ -1 + income + temp', icecream_data)

linear_result3 = linear_model3.fit()


### Viewing Recreated Model Summary 2

In [151]:
print(linear_result3.summary())

                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.990
Method:                 Least Squares   F-statistic:                     1426.
Date:                Tue, 19 Jun 2018   Prob (F-statistic):           6.77e-29
Time:                        12:04:10   Log-Likelihood:                 57.146
No. Observations:                  30   AIC:                            -110.3
Df Residuals:                      28   BIC:                            -107.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
income         0.0023      0.000      9.906      0.0

### Analyzing Model 2

R-squared value increased to **0.990**, which suggests the model could best describe the relationship of **cons** with **income** and **temp** variables.

## HandsOn

Perform the following tasks:

Load the R data set **mtcars**.

Capture the data as **pandas data frame**.

Build a linear regression model with independent variable **wt** and dependent variable **mpg**.

**Fit** the model with **data** and display **R-squared** value.

In [152]:
import statsmodels.api as sm
import pandas as pd
import statsmodels.formula.api as smf

mtcars=sm.datasets.get_rdataset('mtcars')
data=pd.DataFrame(mtcars.data)
liner_model=smf.ols('wt ~  mpg',data)
liner_result=liner_model.fit()
print(liner_result.rsquared)

0.7528327936582647


Load the R data set **mtcars** as pandas dataframe.

Build another linear regression model by considering **log** of independent variable **wt** and **log** of dependent variable **mpg**.

**Fit** the model with data and display **R-squared** value.

In [153]:
import statsmodels.api as sm
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

mtcars=sm.datasets.get_rdataset('mtcars')
data=pd.DataFrame(mtcars.data)
liner_model=smf.ols('np.log(wt) ~  np.log(mpg)',data)
liner_result=liner_model.fit()
print(liner_result.rsquared)

0.8056486518918404


# Logistic Regression with statsmodels

## Discrete Regression - Introduction

A discrete dependent variable takes **few possible outcome** values and is **not normally distributed**.

Hence,** linear regression** cannot be applied to a **discrete variable**.

## Discrete Regression models


*   **statsmodels** provide the following classes to work with discrete regression problems.
    *  **Logit**: for Logistic Regression
    *   **MNLogit**: for Multinomial Logistic Regression
    *   **Poisson**: for Poisson Regression
*   you will see an example of performing Logistic regression and Poisson regression.




## Logistic Regression

Now, let's understand how to perform logistic regression using statsmodels with the following steps:


*   Download the popular **iris** data set (containing data of 3 species) from **R repository**.
*   **Subset** the data of only two species.
*   Perform **transformations**, if required.
*   Define a **patsy** formula and **create a model **using logit.
*   **Fit** the model with supplied data.
*   **View summary** of the model.



### Loading iris Dataset


In [154]:
import statsmodels.formula.api as smf

#Retrieving dataset as pandas data frame.
iris = sm.datasets.get_rdataset("iris").data

#Viewing no. of species
print(iris.Species.unique())

['setosa' 'versicolor' 'virginica']


### Selecting Required Data

As seen in the previous example, **iris** dataset contains details of three types of species.

In a logistic regression, the response variable refers to **only two type** of categories. Hence, data of **two species is filtered** and considered in further steps.

In [155]:
iris_subset = iris[(iris.Species == "versicolor") | (iris.Species == "virginica")].copy()

print(iris_subset.Species.unique())

['versicolor' 'virginica']


### Applying Transformations

A binary variable corresponding to two species is created using map function as shown below.

In [0]:
iris_subset.Species = iris_subset.Species.map({"versicolor": 1, "virginica": 0})

**period** characters, present in column names are replaced with **underscore** characters.

In [0]:
iris_subset.rename(columns={"Sepal.Length": "Sepal_Length",
    "Sepal.Width": "Sepal_Width",
    "Petal.Length": "Petal_Length",
    "Petal.Width": "Petal_Width"}, 
    inplace=True)

### Creating the Model

A logistic regression model, explaining relationship of **Species** variable with **Petal_Length** and **Petal_width** is generated using logit function.

The appropriate patsy formula is passed as argument to **logit** function.

In [0]:
logit_model = smf.logit('Species ~ Petal_Length + Petal_Width', iris_subset)

Model is then fit with supplied data using **fit** function

In [159]:
logit_result = logit_model.fit()

Optimization terminated successfully.
         Current function value: 0.102818
         Iterations 10


### Viewing the Summary

In [160]:
print(logit_result.summary())

                           Logit Regression Results                           
Dep. Variable:                Species   No. Observations:                  100
Model:                          Logit   Df Residuals:                       97
Method:                           MLE   Df Model:                            2
Date:                Tue, 19 Jun 2018   Pseudo R-squ.:                  0.8517
Time:                        12:04:20   Log-Likelihood:                -10.282
converged:                       True   LL-Null:                       -69.315
                                        LLR p-value:                 2.303e-26
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       45.2723     13.612      3.326      0.001      18.594      71.951
Petal_Length    -5.7545      2.306     -2.496      0.013     -10.274      -1.235
Petal_Width    -10.4467      3.756     -2.78

### Analyzing the Model

The model summary indicates that both **Petal_Length** and **Petal_Width** are statistically significant.

For every unit change in **Petal_Length**, the log odds of being **versicolor** decreases by **5.75**.

For every unit change in **Petal_Length**, the log odds of being **versicolor** decreases by **10.44**.

Once satisfied with the model, it can be used to predict response variable value for new values of independent variables.

### Predicting Response Variable

Now, let's create **20** observations having **random** values of **Petal_Length** and **Petal_Width**.

Use these 20 observations to predict values for **Species**.

In [161]:
import pandas as pd
import numpy as np

new = pd.DataFrame({"Petal_Length": np.random.randn(20)*0.5 + 5, "Petal_Width": np.random.randn(20)*0.5 + 1.7})

new["P-Species"] = logit_result.predict(new)

print(new["P-Species"].head())

0    0.975473
1    0.000013
2    0.187547
3    0.239311
4    0.016666
Name: P-Species, dtype: float64


The result is an array of probabilities of response variable being 1.

## HandsOn

Perform the following tasks:

Load the R dataset **biopsy** from **MASS** package.

Capture the data as **pandas data frame**.

Rename the column name **class** to **Class**.

Transform Class column values **benign** and  **malignant** to **0 and 1** respectively.

Build a **logistic regression mode**l with independent variable** V1** and dependent variable **Class**.

**Fit** the model with data and **display Pseudo R-squared** value.

In [162]:
import statsmodels.api as sm
import pandas as pd
import statsmodels.formula.api as smf

biopsy=sm.datasets.get_rdataset('biopsy','MASS')
data=pd.DataFrame(biopsy.data)
data.rename(columns={"class": "Class"}, 
    inplace=True)
data.Class=data.Class.map({'benign':0,'malignant':1})
log_model=smf.logit('Class ~ V1',data)
log_result=log_model.fit()
print(log_result.prsquared)

Optimization terminated successfully.
         Current function value: 0.331941
         Iterations 7
0.48468648480416887


# Poisson Regression with StatsModels

## Poisson Model Example

**Poisson Model** describes a process where dependent variable refers to success count of many attempts and each attempt has a very low probability of success.

Let's understand how to fit a Poisson regression model for a data set available at UCLA repository.

The dataset contains details of a number of awards earned, type of program enrolled, and score obtained in final math exam by students at a high school.

### Loading the Dataset

In [163]:
import pandas as pd

awards_df = pd.read_csv("https://stats.idre.ucla.edu/stat/data/poisson_sim.csv")

print(awards_df.head(3))

    id  num_awards  prog  math
0   45           0     3    41
1  108           0     1    41
2   15           0     3    44


There are three type of programs enrolled by a student : **1 - "General", 2 - "Academic", 3 - "Vocational"**

### Creating the Model

Now let's create a Poisson model with the patsy formula **num_awards ~ math + C(prog).**

In [0]:
import statsmodels.formula.api as smf

poisson_model = smf.poisson('num_awards ~ math + C(prog)', awards_df)

**Fitting the Model**

In [165]:
poisson_model_result = poisson_model.fit()

Optimization terminated successfully.
         Current function value: 0.913761
         Iterations 7


### Viewing Model Summary

In [166]:
print(poisson_model_result.summary())

                          Poisson Regression Results                          
Dep. Variable:             num_awards   No. Observations:                  200
Model:                        Poisson   Df Residuals:                      196
Method:                           MLE   Df Model:                            3
Date:                Tue, 19 Jun 2018   Pseudo R-squ.:                  0.2118
Time:                        12:04:27   Log-Likelihood:                -182.75
converged:                       True   LL-Null:                       -231.86
                                        LLR p-value:                 3.747e-21
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -5.2471      0.658     -7.969      0.000      -6.538      -3.957
C(prog)[T.2]     1.0839      0.358      3.025      0.002       0.382       1.786
C(prog)[T.3]     0.3698      0.441      0.83

### Analyzing Model Summary

The coefficient for **math** variable is** 0.07**, which means for every one unit increase in **math**, the log count increases by 0.07.

Having enrolled for** prog=2, i.e., "Academic"**, instead of "Generic" program, changes the log count by 1.08.

Having enrolled for **prog=3, i.e., "Vocational"**, instead of "Generic" program, changes the log count by 0.37.

## HandsOn

Perform the following tasks:

Load the R data set **Insurance** from **MASS** package.

Capture the data as **pandas data frame**.

Build a Poisson regression model with a **log** of an independent variable, **Holders** and dependent variable **Claims**.

**Fit** the model with data and find the **sum of residuals**.

In [167]:
import statsmodels.api as sm
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

insurance=sm.datasets.get_rdataset('Insurance','MASS')
data=pd.DataFrame(insurance.data)
pos_model=smf.poisson(' Claims ~ np.log(Holders)',data)
pos_result=pos_model.fit()
print(sum(pos_result.resid))

         Current function value: 2438780692723310424842348299382319651840164623859046154240.000000
         Iterations: 35
-1.5608196433429187e+59


/usr/local/lib/python3.6/dist-packages/statsmodels/base/model.py:496: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)


# ANOVA with statsmodels

Let's now understand how to perform ANOVA of a fitted regression model.

For this example, let's once again consider popular **Icecream** dataset, used in **linear regression** section.

## Loading the Dataset

In [0]:
import statsmodels.api as sm

icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")
icecream_data = icecream.data

## Building the Model

In [0]:
import statsmodels.formula.api as smf

model1 = smf.ols('cons ~ temp', icecream_data).fit()

A **linear regression model **of icecream consumption with temperature is built using **ols** method.

Now let's frame a null hypothesis: Value of **B1**, i.e., the coefficient of temp variable is zero.

## Creating ANOVA Table

**anova_lm** method from **anova** module of **statsmodels** library can be used for performing **ANOVA**.

In [170]:
from statsmodels.stats import anova
print(anova.anova_lm(model1))

            df    sum_sq   mean_sq         F        PR(>F)
temp       1.0  0.075514  0.075514  42.27997  4.789215e-07
Residual  28.0  0.050009  0.001786       NaN           NaN


/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


The obtained **F-statistic **is **42.27** and has a **very low probability.**

Hence, the null hypothesis can be rejected. i.e., B1 is not equal to zero

## Building New Model

Now let's create a new model with two independent variables and one response variable.

With more that one independent variable, the null hypothesis is stated as Coefficients of all independent variables are zero. i.e** B1 = 0, and B2 = 0.**

The alternative hypothesis will be that atleast one of the parameters** Bj != 0** where **j takes the values 1, 2, ..**.

In [0]:
model2 = smf.ols('cons ~ income + temp', icecream_data).fit()

## Creating ANOVA Table 1

In [172]:
print(anova.anova_lm(model2))

            df    sum_sq   mean_sq          F        PR(>F)
income     1.0  0.000288  0.000288   0.208231  6.518069e-01
temp       1.0  0.087836  0.087836  63.413711  1.470071e-08
Residual  27.0  0.037399  0.001385        NaN           NaN


/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


The above table shows **F-statistic** corresponding to each independent variable.

## Determining F-Value

**F-statistic** is defined as **Mean square of model/Mean square of residuals**.

Mean square of model is** sum of squares values of all variables/degrees of freedom** i.e (0.000288 + 0.087836)/2 = 0.044062.

Hence, the** F-statistic** of model is **(0.044062 / 0.001385) = 31.813.**

The probability of obtained F-statistic is 7.96*e-08 and is computed as shown below:

In [173]:
print(stats.f.sf(31.81, 2, 27))

7.959504548627583e-08


Since the **p-value** is **low**,** null hypothesis is rejected**.

## Comparing Two Models

ANOVA can also be used to compare two nested or related models.

**model1** regresses consumption with temperature and **model2** with temperature and income.

Below code verifies if the decrease in residuals sum of squares is significant or not.

In [174]:
print(anova.anova_lm(model1, model2))

   df_resid       ssr  df_diff   ss_diff         F    Pr(>F)
0      28.0  0.050009      0.0       NaN       NaN       NaN
1      27.0  0.037399      1.0  0.012611  9.104375  0.005506


/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


**p-value i.e 0.005** from above table suggests the decrease in residual sum of squares in **model2**, compared to **model1** is significant.

## HansOn

Perform **ANOVA** on the **first linear model** obtained while working with **mtcars** data set.

Display the **F-statistic** value.

In [175]:
import statsmodels.api as sm
import pandas as pd
import statsmodels.formula.api as smf
from statsmodels.stats import anova

mtcars=sm.datasets.get_rdataset('mtcars')
data=pd.DataFrame(mtcars.data)
liner_model=smf.ols('wt ~  mpg',data)
liner_result=liner_model.fit()
ano=anova.anova_lm(liner_result)
print(ano.F[0])
#print(stats.f.sf(91.375325, 1, 30))

91.37532500376184


/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


Perform **ANOVA** on the **second linear model** obtained while working with **mtcars** data set.

Display the **F-statistic** value.

In [176]:
import statsmodels.api as sm
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from statsmodels.stats import anova

mtcars=sm.datasets.get_rdataset('mtcars')
data=pd.DataFrame(mtcars.data)
liner_model=smf.ols('np.log(wt) ~  np.log(mpg)',data)
liner_result=liner_model.fit()
ano=anova.anova_lm(liner_result)
print(ano.F[0])

124.35961876273966


/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1818: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)
